### ISSUE TITLE
Epic: single source of truth for document/project information

### ISSUE THREAD

**Issue by cscheid** [quarto-dev/quarto-cli#8700]
A number of bugs in Quarto happen because different parts of the code are responsible for reasoning about a .qmd document. We should have a single source of truth on this, and that should be what drives the output of, for example,`quarto inspect`.

This is related (but simpler) to our Execution Planner, and solving this will make it easier to do the planner well.

- [ ] #3185 
- [ ] #3157 
- [ ] #3053 
- [ ] #2216

**Comment by jkrumbiegel** [comment]
I see there's been some work in this direction, with the addition of the `ProjectContext` in https://github.com/quarto-dev/quarto-cli/pull/8771. For our users, both not being able to specify `engine: julia` in the `_quarto.yml` and not being able to use `.jl` script files specifying `engine: julia` are still annoying roadblocks that I'd like to remove. @cscheid could you tell me if you think this can be meaningfully done by an outside contributor, or are the necessary changes too intertwined with your larger refactoring plans for the code base? If it's possible, it would be great to get a rough idea from you which parts are missing.

I was looking at the `fileExecutionEngine` function. In principle, the logic seems relatively straightforward. If an engine is specified in either frontmatter or project, this should take precedence over "file claiming". An engine specified in the frontmatter should take precedence over one specified in the project. With your new `ProjectContext`, the information needed is already there. However, `markdownExecutionEngine` currently expects file claiming to have run before it, and I'm not sure which way this conflict should be resolved.

**Comment by cderv** [comment]
> not being able to specify engine: julia in the _quarto.yml

For reference, project wide engine setting is tracked at: 
- https://github.com/quarto-dev/quarto-cli/issues/3157

This is a limitation across engine currently I believe. 

> being able to use .jl script files specifying engine: julia

For script file, we started discussion at 
- https://github.com/quarto-dev/quarto-cli/issues/10034

Is this the same topic ? 

I am just cross-referencing to keep track of related topics so that we can all follow.

**Comment by jkrumbiegel** [comment]
Yes, thanks, should've just linked those again directly.

**Comment by cscheid** [comment]
@jkrumbiegel Yes, I definitely want to fix this. We're nearing a 1.6 release and won't have time to do this between now and then (we plan to release early November). But I would like to spend some time on this in 1.7. Could we revisit this then?

I hope to be able to split the work such that some of it can be tackled by the engine.

**Comment by jkrumbiegel** [comment]
Ok sure, if there's nothing I could do now, let's revisit it in November. But I could also get started now if I had a rough outline how to approach these two issues. Whatever works better for you :)

**Comment by cscheid** [comment]
> For our users, both not being able to specify `engine: julia` in the `_quarto.yml`

I think this one will be harder for you to fix, because it involves the code where we bind engines/"targets" to documents. It is pretty messy and mostly outside of an engine's responsibilities.

> and not being able to use `.jl` script files specifying `engine: julia` are still annoying roadblocks that I'd like to remove.

I think this could be easier (barring unexpected snags in our codebase). I think this one is entirely a matter of implementing `claimsFile` in the Julia engine. For contrast with the `knitr` and `jupyter` engines, see:

https://github.com/quarto-dev/quarto-cli/blob/8eac0b5cf371ac26c0f0584dae5c00f572769210/src/execute/rmd.ts#L68-L71

https://github.com/quarto-dev/quarto-cli/blob/0097478870de7df06794b523eb49e5dcdcf6a771/src/execute/jupyter/jupyter.ts#L148-L151

While the julia engine has:

https://github.com/quarto-dev/quarto-cli/blob/9f4e18803d421cf0fb7a224b4cfc52b104fb8e5a/src/execute/julia.ts#L70

When an engine claims a file, then it's also responsible for providing the converted markdown for that file. See eg the jupyter engine code:

https://github.com/quarto-dev/quarto-cli/blob/0097478870de7df06794b523eb49e5dcdcf6a771/src/execute/jupyter/jupyter.ts#L162-L165

**Comment by jkrumbiegel** [comment]
Hm are you saying I should check within claimsFile whether the percent script has a `engine: julia` setting? Because otherwise I can't claim jl files without it being a breaking change. That's why I didn't do it at the time.

**Comment by cscheid** [comment]
> Because otherwise I can't claim jl files without it being a breaking change. That's why I didn't do it at the time.

I only now see that jupyter.ts claims `.jl` percent scripts. That does complicate things quite a bit. I now think that this is _also_ a November conversation. Should we put something on the calendar so that we can have a discussion about it?

**Comment by jkrumbiegel** [comment]
That would be nice. Should I send you an email?

**Comment by cscheid** [comment]
Yes, please: carlos.scheidegger@posit.co
